core/ecschema-rpc/impl/src/ECSchemaRpcImpl.ts

```diff
@@ -0,0 +1,95 @@
+import { SchemaKey } from "@bentley/ecschema-metadata";
+import { ClientRequestContext } from "@bentley/bentleyjs-core";
+
+/**
+ * Defines the interface how the rows of the iModel query look like.
+ */
+interface SchemaNameRow {
+  schemaName: string;
+  read: string;
+  write: string;
+  minor: string;
+}
+
+/**
+ * Implementation of the SchemaRpcInterface.
+ * @internal
+ */
+export class ECSchemaRpcImpl extends ECSchemaRpcInterface {
+  /**
+   * Registers the RPC interface with its corresponding implementation class.
+   */
+  public static register() {
+    RpcManager.registerImpl(ECSchemaRpcInterface, ECSchemaRpcImpl);
+  }
+
+  /**
+   * Gets an iModelDb instance. It is important that the database has been opened before
+   * otherwise it can't be found.
+   * @param tokenProps        The iModelToken props that hold the information which iModel is used.
+   * @returns                 Instance of IModelDb.
+   */
+  private async getIModelDatabase(tokenProps: IModelRpcProps): Promise<backend.IModelDb> {
+    return new Promise<backend.IModelDb>((resolve) => {
+      resolve(backend.IModelDb.findByKey(tokenProps.key));
+    });
+  }
+
+  /**
+   * Returns an array of SchemaKey that exists in the current iModel context.
+   * @param tokenProps        The iModelToken props that hold the information which iModel is used.
+   * @returns                 An array of SchemaKey.
+   */
+  public async getSchemaKeys(tokenProps: IModelRpcProps): Promise<SchemaKey[]> {
+    ClientRequestContext.current.enter();
+
+    const schemaKeys: SchemaKey[] = [];
+    const iModelDb = await this.getIModelDatabase(tokenProps);
+
+    // Iterate over the rows returned from AsyncIterableIterator. The custom Query overload returns
+    // a typed row instance instead of any.
+    const schemaNameQuery = `SELECT Name as schemaName, VersionMajor as read, VersionWrite as write, VersionMinor as minor FROM main.meta.ECSchemaDef`;
+    for await (const schemaDefinitionRow of iModelDb.query(schemaNameQuery) as AsyncIterableIterator<SchemaNameRow>) {
+      const schemaFullName = schemaDefinitionRow.schemaName;
+      const read = Number(schemaDefinitionRow.read);
+      const write = Number(schemaDefinitionRow.write);
+      const minor = Number(schemaDefinitionRow.minor);
+      schemaKeys.push(new SchemaKey(schemaFullName, read, write, minor));
+    }
+    return schemaKeys;
+  }
+
+  /**
+   * Gets the schema JSON for the current iModel context and returns the schema as a string which the client can parse to SchemaProps.
+   * @param tokenProps        The iModelToken props that hold the information which iModel is used.
+   * @param schemaName        The name of the schema that shall be returned.
+   * @returns                 The SchemaProps as a string.
+   */
+  public async getSchemaJSON(tokenProps: IModelRpcProps, schemaName: string): Promise<string> {
+    ClientRequestContext.current.enter();
+
+    if (schemaName === undefined || schemaName.length < 1) {
```

Comment:
If the schema name can be potentially undefined, then we should probably change the type expected.

It would be interesting to call the backend without it defined in this way, to see what would actually happen.  I'm pretty sure it would get through to this line but not certain about that.

Reply:
We don't want it to be undefined ... we want it to always be set ... but can we really assume it will always be defined  when coming over the wire?  Isn't it safer to check?

I think the intent is clearer with the schemaName being required

Reply:
Can you try it to see if that's the case?  If it does allow `undefined` through then that's fine to me to check but right now the code is confusing since the type says something different than the checks.

Reply:
The schema name is required, so the API signature should make that clear.  If it is not possible for it to be undefined then the undefined check should be removed instead of changing the API signature to allow undefined.

Reply:
I'll make the change.